### PR TITLE
Add .tvips file reader plug-in

### PR DESCRIPTION
**Pull request by din14970**
### Description of the change
As discussed in #2780 this PR aims to add reading capabilities for .tvips files.

It needs more testing but at first sight it seems to work on most of the files I have.  It will probably take a bit longer to finish this as I will need to create dummy files for the tests - the real files I have are all huge.

### Progress of the PR
- [x] Change implemented
  - [x] `file_reader` function that parses the data and metadata
  - [x] binding to the `io.save` function
- [x] update docstring (if appropriate),
- [x] update user guide (if appropriate),
- [x] add an changelog entry in the `upcoming_changes` folder (see [`upcoming_changes/README.rst`](https://github.com/hyperspy/hyperspy/blob/RELEASE_next_minor/upcoming_changes/README.rst)),
- [x] Check formatting changelog entry in the `readthedocs` doc build of this PR (link in github checks)
- [x] add tests,
- [x] ready for review.

### Minimal example of the bug fix or the new feature
```python
import hyperspy.api as hs
# loads as an image stack
hs.load("datafile.tvips", lazy=True)
# loads as a 4D STEM dataset
hs.load("datafile.tvips", lazy=True, scan_shape=(100, 100), scan_start_frame=20)
```

### CODE REVIEW

File: hyperspy/io_plugins/tvips.py
Comment (harripj) [inline hyperspy/io_plugins/tvips.py]: Should this be ```'tvips'``` and ```'TVIPS'```?

**Review by harripj — CHANGES_REQUESTED** [review]
Requesting clarity on file_extensions for the plugin.

File: hyperspy/io_plugins/tvips.py
Comment (din14970) [reply hyperspy/io_plugins/tvips.py]: yes, I haven't had time to look at this recently, I think at the beginning of october I can resume work on this

File: hyperspy/io_plugins/tvips.py
Comment (ericpre) [inline hyperspy/io_plugins/tvips.py]: Following up on discussion in https://github.com/din14970/TVIPSconverter/issues/15#issuecomment-916041983, in these dictionary, a value should be defined for the `'navigate'` key to set the navigation axes (`'navigate'=True` means that this axis is a navigation axes!).

File: hyperspy/io_plugins/tvips.py
Comment (harripj) [inline hyperspy/io_plugins/tvips.py]: ```scan_shape``` could be adapted to take any number of dimensions, for example to load the signal as 1d. Something like this would seem to work for any dimensionality:
```python
total_scan_frames = np.prod(scan_shape)
indices = np.arange(scan_start_frame, scan_start_frame+total_scan_frames).reshape(scan_shape)
```

**Review by harripj — CHANGES_REQUESTED** [review]

File: hyperspy/io_plugins/tvips.py
Comment (harripj) [reply hyperspy/io_plugins/tvips.py]: This addition fixed things for me:
```'navigate': True if i < len(scan_shape) else False```

File: hyperspy/io_plugins/tvips.py
Comment (harripj) [inline hyperspy/io_plugins/tvips.py:418]: Following on from the previous [comment ](https://github.com/hyperspy/hyperspy/pull/2781/files/a1bee7caf10dcac17a9785eb38a715bf594e3dc2#r705321351), a similar addition here:
```'navigate': True if not i else False```

File: doc/user_guide/io.rst
Comment (ericpre) [inline doc/user_guide/io.rst]: Specify default value for `winding_scan_axis` and `hysteresis`.

File: doc/user_guide/io.rst
Comment (ericpre) [inline doc/user_guide/io.rst:1977]: Specify default value for `max_file_size` and `frame_header_extra_bytes`.

File: hyperspy/io_plugins/tvips.py
Comment (ericpre) [inline hyperspy/io_plugins/tvips.py]: Use `from hyperspy.api_nogui import _ureg` instead.

As per https://pint.readthedocs.io/en/stable/tutorial.html#using-pint-in-your-projects.

File: hyperspy/io_plugins/tvips.py
Comment (ericpre) [inline hyperspy/io_plugins/tvips.py]: Specify what it does when it is `None`.

File: hyperspy/io_plugins/tvips.py
Comment (ericpre) [inline hyperspy/io_plugins/tvips.py]: Is there any value in making rechunking optional (possibly with a default to `True`, if this is sensible) and having it as a parameter of the reader? If so, there is another one in the `else` branch below.

File: hyperspy/io_plugins/tvips.py
Comment (ericpre) [inline hyperspy/io_plugins/tvips.py]:
```suggestion
                show_progressbar = kwds.get(
                    'show_progressbar',
                    preferences.General.show_progressbar
                    )
```
And remove the line setting `show_progressbar` above

**Comment by din14970** [comment]
@ericpre @harripj sorry for the long delay, I think this should be good to go

**Comment by harripj** [comment]
@din14970 no need to be sorry, thank you for your work!

I will preface my comments here by acknowledging that I am by no means an expert in hyperspy, so please forgive me if I am making some basic mistakes.

Firstly this is an excellent addition to hyperspy and very useful for anybody working with data from TVIPS cameras and data.

I have tested this on some .tvips data that we have collected (not in the testing data uploaded to this PR). On my Mac the `lazy` loader works well and does not load the whole file to memory. We discussed [previously](https://github.com/din14970/TVIPSconverter/issues/15#issuecomment-915967600) that there may be an issue on Windows, but if I understood, this is actually related to a Dask issue and not this PR.

When I run the example, eg. `hs.load('datafile.tvips', lazy=True)`, the following warning is always raised for me: ```WARNING:hyperspy.io:`signal_type='imaging'` not understood. See `hs.print_known_signal_types()` for a list of installed signal types or https://github.com/hyperspy/hyperspy-extensions-list for the list of all hyperspy extensions providing signals.``` Perhaps there is an extension that I need to install to suppress this warning? I then installed `pyxem` but the warning persists. Is the `"imaging"` signal type related to another package?

If I try to load a file as 1d or 2d, all works as expected. If the `scan_shape` provided is too large for the file the following error is raised: `IndexError: Index is out of bounds for axis 0 with size 19143`, but perhaps this could be more informative. It would be nice if the number of frames in the dataset(s) can be determined beforehand to determine if the scan indices (shape + offset) fit and a clearer error is raised to the user.

To me this looks great, thanks very much for working on this! I will leave this to someone involved with the hyperspy project to provide some more feedback.

**Comment by din14970** [comment]
@harripj Thanks for having another look. The goal would of course be to make it work on real TVIPS datasets, the test datasets were actually artificially created.

> We discussed https://github.com/din14970/TVIPSconverter/issues/15#issuecomment-915967600 that there may be an issue on Windows, but if I understood, this is actually related to a Dask issue and not this PR.

Yes I'm not sure how to deal with this, I also don't have a Windows machine on which I could test/debug. Perhaps if it's a real issue for you you could have a deeper look into the task graph on Windows to figure out what is going on?

> WARNING:hyperspy.io:`signal_type='imaging'` not understood

You actually caught a bug here, so thanks for that - I use `signal_type` in the metadata to distinguish between diffraction and imaging data, but it seems hyperspy uses this info for more than just information purposes. Will remove this.

>  It would be nice if the number of frames in the dataset(s) can be determined beforehand to determine if the scan indices (shape + offset) fit and a clearer error is raised to the user.

Good idea, will implement

**Comment by harripj** [comment]
> The goal would of course be to make it work on real TVIPS datasets, the test datasets were actually artificially created.

For our experimental datasets it works as expected!

> Yes I'm not sure how to deal with this, I also don't have a Windows machine on which I could test/debug. Perhaps if it's a real issue for you you could have a deeper look into the task graph on Windows to figure out what is going on?

I can test this PR again on Windows next week and let you know. I think even if the problem exists it is likely Dask related and beyond the scope of this PR. I don't think that this potential issue should stop this PR from being merged.

Thanks again for implementing this @din14970!

File: hyperspy/io_plugins/tvips.py
Comment (ericpre) [inline hyperspy/io_plugins/tvips.py]: You can use any signal type listed in https://github.com/hyperspy/hyperspy-extensions-list. In this case, it could be `diffraction` or `electron_diffraction`. If I remember correctly, TVIPS camera can be used for X-rays, so if there is way to make sure that it was an electron diffraction, then it would be more accurate to use `electron_diffraction`. 

For imaging case, it should be left "empty" as there is no extension which implements a corresponding signal.

**Review by ericpre — COMMENTED** [review]
This looks very good! I added a few suggestions.

It would be good to try to get the full coverage - see codecov warning in the "files changed" tab. When it is not possible to cover or not worth it, you can use `# pragma: no cover`.

**Comment by ericpre** [comment]
> > Yes I'm not sure how to deal with this, I also don't have a Windows machine on which I could test/debug. Perhaps if it's a real issue for you you could have a deeper look into the task graph on Windows to figure out what is going on?
> 
> I can test this PR again on Windows next week and let you know. I think even if the problem exists it is likely Dask related and beyond the scope of this PR. I don't think that this potential issue should stop this PR from being merged.

That would be great! Let see how it goes and then think about it if we need to! ;)

File: doc/user_guide/io.rst
Comment (din14970) [reply doc/user_guide/io.rst]: Done

File: doc/user_guide/io.rst
Comment (din14970) [reply doc/user_guide/io.rst:1977]: done

File: hyperspy/io_plugins/tvips.py
Comment (din14970) [reply hyperspy/io_plugins/tvips.py]: done

File: hyperspy/io_plugins/tvips.py
Comment (din14970) [reply hyperspy/io_plugins/tvips.py]: done

File: hyperspy/io_plugins/tvips.py
Comment (din14970) [reply hyperspy/io_plugins/tvips.py]: I now only add the `Signal` key if mode is 2 and set `sigtype` to `diffraction`. I don't know whether it's possible to distinguish between electrons or X-rays, but it should be relatively easy for users to change from diffraction to electrondiffraction if relevant.

File: hyperspy/io_plugins/tvips.py
Comment (din14970) [reply hyperspy/io_plugins/tvips.py]: done

File: hyperspy/io_plugins/tvips.py
Comment (din14970) [reply hyperspy/io_plugins/tvips.py]: I've made `rechunking` a kwarg. If it's "falsy" no rechunking occurs. I'm pretty sure this would mean each file corresponds to a chunk. If it's "auto" = default then optimal rechunking occurs along navigation dimensions as above. If it's anything else, it's passed directly to the `chunks` argument in `dask.array.rechunk`, in case the user wants to do a custom rechunking.

**Comment by din14970** [comment]
@ericpre A couple lines are missing from coverage because I don't have files to test them on. I only ever found and collected data with header v2. It's in there because I found it in an older implementation for a reader. It is possible the format evolves further, which the exceptions should take care of. I've added the pragma lines as you suggest.

On a subject totally unrelated to this PR, it appears your affiliation has changed and your e-mail address is no longer valid. I've sent e-mails to your superstem account but I'm not sure if this one is valid either. If not, you probably haven't received any of the e-mail traffic related to the pyxem paper on which I invited you as co-author, so I'm informing you via this channel that the paper is now accepted in Ultramicroscopy. Could you send me an e-mail if you want to review the proofs once I receive them?

**Review by ericpre — APPROVED** [review]
Looks good to me. Before merging, it would be good to know the outcome of @harripj's testing on windows. Thanks!

**Comment by harripj** [comment]
Just a follow up on this. On my Windows machine I seem to be experiencing the same behaviour as [before](https://github.com/din14970/TVIPSconverter/issues/15#issuecomment-915967600):

If I load one of our experimental datasets (size 9.9 GB) as ` data = hs.load('dset.tvips', lazy=True)` the file is not loaded to memory as expected. With `lazy=False` the peak memory usage is double the file size before settling back to ~file size:

image

Using the lazy `loader` again, if I just extract the center pixel from each frame as:
```python
center = data[:, 256, 256].compute()
```
The whole file is loaded to memory and persists:
image

Interestingly the memory presists even if I call `del center; del data`.

Dask version:  `2022.03.0`
NumPy version : `1.21.5`
Hyperspy version: `1.7.0.dev0`, ie. local install of this PR.

**Comment by ericpre** [comment]
Thanks @harripj for checking it. Let's merge this PR and park it as a "known issue on windows"... Fixing it can be done in a separate PR.

**Pull request merged**
